### PR TITLE
Preload API requests

### LINKED ISSUE #499: Preload REST API calls
## Feature Description
Currently when the editor loads, it calls the REST API to get data. 

Screenshot 2020-03-06 at 13 44 26

These calls are 
- Load media ( Video and images ) for library.
- Load post ( story ) from editor. 
- Load list of fonts to display fonts. 

When the documents panel is active, two more calls are made. 

Screenshot 2020-03-06 at 13 47 12

These calls are 
- Load a list of Post statuses,
- Load of list of users. 

We currently load from data into the editor via config which is a [js variable](https://github.com/google/web-stories-wp/blob/c29f388b8135b9b20e061d036c01222f8fa10cbb/includes/Story_Post_Type.php#L301-L323) on the page. This could be expanded to include some or all data listed above. 

Or we could use `rest_preload_api_request`, which preloads from REST API [before the editor](https://github.com/WordPress/wordpress-develop/blob/437ac63d7652b08095dc7f77b6ba4ef52db44dbb/src/wp-admin/edit-form-blocks.php#L41-L87) is loaded. 

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria
- What data should be preloaded 
- How should it be preloaded

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Use [ApiFetch](https://github.com/WordPress/gutenberg/tree/9b6fee52af1066294e4212c0affd87dfb197bc9b/packages/api-fetch) preload middleware functionality to preload data in the editor. 
This is similar to the gutenberg editor preloads [data](https://github.com/WordPress/wordpress-develop/blob/437ac63d7652b08095dc7f77b6ba4ef52db44dbb/src/wp-admin/edit-form-blocks.php#L41-L87). 

## Relevant Technical Choices
This change is a very limited at this time. Only the current preloads the fonts api and current story endpoint. 

Other requests, research to preload. 

### Story editor
- `wp/v2/media?context=edit&page=1&per_page=100`
- `wp/v2/users?per_page=100`

### Dasboard
- `wp/v2/web-story?context=edit&order=desc&orderby=modified&page=1&per_page=20&status=publish%2Cdraft`
- `/wp/v2/web-story?context=edit&order=desc&orderby=modified&page=1&per_page=20&status=publish`
- `wp/v2/web-story?context=edit&order=desc&orderby=modified&page=1&per_page=20&status=draft`
- `wp/v2/users?per_page=100`

Media and filtered requests for web stories, all pass, `parse: false` ( see this [example](https://github.com/google/web-stories-wp/blob/9669b33a1f4bbf7f72ee5a8866aefc79e7fe0c5a/assets/src/edit-story/app/api/apiProvider.js#L126)), meaning they can not be preloaded. See this line [code](https://github.com/WordPress/gutenberg/blob/9b6fee52af1066294e4212c0affd87dfb197bc9b/packages/api-fetch/src/middlewares/preloading.js#L48). 

Similar with request to users, this has the param `per_page: -1` ( [example](https://github.com/google/web-stories-wp/blob/9669b33a1f4bbf7f72ee5a8866aefc79e7fe0c5a/assets/src/edit-story/app/api/apiProvider.js#L226  )). This too is passes  `parse: false` in the [fetch-all middleware](https://github.com/WordPress/gutenberg/blob/9b6fee52af1066294e4212c0affd87dfb197bc9b/packages/api-fetch/src/middlewares/fetch-all-middleware.js#L61-L67).

## To-do
- Templates, can be preloaded via the API.
- Users / Media would require refactoring to preload and this would be up to those pods to look into. 

## User-facing changes

N / A

## Testing Instructions

This should not be a user facing change. 

---
Fixes #499

### CODE REVIEW

File: includes/Dashboard.php [hunk 1]
```diff
@@ -85,6 +86,40 @@ public function add_menu_page() {
...
+		];
+
+		/**
+		 * Preload common data by specifying an array of REST API paths that will be preloaded.
+		 *
+		 * Filters the array of paths that will be preloaded.
+		 *
+		 * @param string[] $preload_paths Array of paths to preload.
+		 */
+		$preload_paths = apply_filters( 'stories_dashboard_preload_paths', $preload_paths );
```
Comment (swissspidy) [inline includes/Dashboard.php]:
```suggestion
		$preload_paths = apply_filters( 'web_stories_dashboard_preload_paths', $preload_paths );
```

File: includes/edit-story.php [hunk 2]
```diff
@@ -29,6 +29,48 @@
...
+
+/**
+ * Preload common data by specifying an array of REST API paths that will be preloaded.
+ *
+ * Filters the array of paths that will be preloaded.
+ *
+ * @param string[] $preload_paths Array of paths to preload.
+ * @param WP_Post  $post          Post being edited.
+ */
+$preload_paths = apply_filters( 'stories_editor_preload_paths', $preload_paths, $post );
```
Comment (swissspidy) [inline includes/edit-story.php]:
```suggestion
$preload_paths = apply_filters( 'web_stories_editor_preload_paths', $preload_paths, $post );
```

Comment (spacedmonkey) [reply includes/edit-story.php]: Fixed https://github.com/google/web-stories-wp/pull/2468/commits/582a04b1a03f0609f8a578b04745fc230d907472

File: includes/Dashboard.php (hunk 1 above)
Comment (spacedmonkey) [reply includes/Dashboard.php]: Fixed https://github.com/google/web-stories-wp/pull/2468/commits/582a04b1a03f0609f8a578b04745fc230d907472

**Review by swissspidy — APPROVED** [review]

**Pull request merged**